### PR TITLE
Assignment3

### PR DESCRIPTION
**Pull request by KS-D**
I plan to refactor my tests a bit in the next day or two. There are a lot tests that are essentially the same for the all the services.

### CODE REVIEW

**Comment by KS-D** [comment]
It looks like I might have rebased off of my master wrong.

**Comment by KS-D** [comment]
Never mind, I fixed it.

File: SecretSanta/src/SecretSanta.Data/Gift.cs [hunk 1]
```diff
@@ -16,5 +14,18 @@ public class Gift : FingerPrintEntityBase
...
+            User = user;
+        }
+#nullable disable
+        private Gift(string title, string description, string url)
+#nullable enable
+        {
+            Title = title;
+            Description = description;
+            Url = url;
+        }
```
Comment (GamerBah) [inline SecretSanta/src/SecretSanta.Data/Gift.cs]: `User` will be set here, but what about `UserId`?

File: SecretSanta/test/SecretSanta.Data.Tests/GiftTests.cs [hunk 2]
```diff
@@ -36,34 +37,131 @@ public async Task Gift_CanBeSavedToDatabase()
...
+            Gift ringGift = SampleData.CreateRingGift();
+            Gift arduino = SampleData.CreateGiftArduino();
+
+            using (var applicationDbContext = new ApplicationDbContext(Options))
+            {
+                applicationDbContext.Gifts.Add(ringGift);
+                await applicationDbContext.SaveChangesAsync();
+                
+                applicationDbContext.Gifts.Add(arduino);
+                await applicationDbContext.SaveChangesAsync();
```
Comment (GamerBah) [inline SecretSanta/test/SecretSanta.Data.Tests/GiftTests.cs]: You might only need to do `await applicationDbContext.SaveChangesAsync();` once after you update/add many objects?

File: SecretSanta/src/SecretSanta.Business/GiftService.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        public override async Task<Gift> FetchByIdAsync(int id)
+        {
+            return await ApplicationDbContext.Gifts.Include(g => g.User)
+                .SingleOrDefaultAsync(item => item.Id == id);
+        }
+
+        public override async Task<List<Gift>> FetchAllAsync()
+        {
+            return await ApplicationDbContext.Gifts.Include(gift => gift.User).ToListAsync();
+        }
```
Comment (GamerBah) [inline SecretSanta/src/SecretSanta.Business/GiftService.cs]: Awesome job overriding the fetch-all. I probably should've thought about that... lol

**Comment by GamerBah** [comment]
# Assignment 3

**Important Note:** Assignment due dates are one day later for Assignment 3

- Connect Azure DevOps up for your fork.
- Add a Status Badge (see [Add a status badge to your repository](https://docs.microsoft.com/en-us/azure/devops/pipelines/create-first-pipeline?view=azure-devops&tabs=browser%2Ctfs-2018-2#add-a-status-badge-to-your-repository)) to the `README.md` file in `master` that shows the build status on each branch. ✔
- Be sure that clicking on the badge successfully navigates one to the Azure DevOps pipeline. ✔
- Rebase to the Readme.md update on `master` to the `Assignment3` branch so it too shows the badge (be sure it points to the Assignment3 branch status) ✔
- Appropriately encapsulate Id and all fingerprint data. ✔
- Add non-default constructors to Gift and address all warnings (see BlogEngine's Post.cs file). ✔
- Add any missing entity framework CRUD unit tests for `Gift`. ✔
- Implement an Automapper Configuration between two `Gift` objects. ✔
  - Do not map the Id ✔
- Create services for (do not map primary key Id for all entities):
  - Groups ✔
  - Users ✔
  - Gifts ✔
  - When retrieving the Gift object from the service, User should be pre-populated (incuded) ✔
- Refactor out SampleData (see BlogEngine's SampleData.cs) ✔

## Extra Credit

The following are options for extra credit (you don't need to do them all):

- Write exceptional code that is clean, conforms to all coding guidelines, is well refactored, and easy to understand. ✔

File: SecretSanta/test/SecretSanta.Data.Tests/GiftTests.cs (hunk 2 above)
Comment (KS-D) [reply SecretSanta/test/SecretSanta.Data.Tests/GiftTests.cs]: Whoops, you are probably right there.

File: SecretSanta/src/SecretSanta.Data/Gift.cs (hunk 1 above)
Comment (KS-D) [reply SecretSanta/src/SecretSanta.Data/Gift.cs]: It probably is a good idea to set UserId here.

File: SecretSanta/test/SecretSanta.Business.Tests/AutomapperProfileConfigurationTests.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    {
+        class MockUser : User
+        {
+            public MockUser(int id, string firstName, string lastName) : base(firstName, lastName)
+            {
+                base.Id = id;
+            }
+        }
+
+        class MockGift : Gift
```
Comment (MarkMichaelis) [inline SecretSanta/test/SecretSanta.Business.Tests/AutomapperProfileConfigurationTests.cs]: AutomapperProfileConfigurationTests.MockGift is an internal class that is apparently never instantiated. If so, remove the code from the assembly. If this class is intended to contain only static members, make it static (Shared in Visual Basic).

File: SecretSanta/src/SecretSanta.Web/Startup.cs [hunk 5]
```diff
@@ -4,11 +4,11 @@
 using Microsoft.Extensions.Hosting;
 
 namespace SecretSanta.Web
-{
-    // Justification: Disable until ConfigureServices is added back.
-#pragma warning disable CA1052 // Static holder types should be Static or NotInheritable
```
Comment (MarkMichaelis) [inline SecretSanta/src/SecretSanta.Web/Startup.cs]: Agreed!

File: SecretSanta/test/SecretSanta.Business.Tests/EntityServiceBaseTest.cs [hunk 6]
```diff
@@ -0,0 +1,151 @@
+﻿using System.Collections.Generic;
```
Comment (MarkMichaelis) [inline SecretSanta/test/SecretSanta.Business.Tests/EntityServiceBaseTest.cs]: Awesome!!!! This is exactly what I would have hoped for. You are the first one to see this.  Great job!!!

**Comment by MarkMichaelis** [comment]
# Assignment 3

## Main Assignment

- Connect Azure DevOps up for your fork.
- Add a Status Badge (see [Add a status badge to your repository](https://docs.microsoft.com/en-us/azure/devops/pipelines/create-first-pipeline?view=azure-devops&tabs=browser%2Ctfs-2018-2#add-a-status-badge-to-your-repository)) to the `README.md` file in `master` that shows the build status on each branch. ✔
- Be sure that clicking on the badge successfully navigates one to the Azure DevOps pipeline. ✔
- Rebase to the Readme.md update on `master` to the `Assignment3` branch so it too shows the badge (be sure it points to the Assignment3 branch status) ✔
- Appropriately encapsulate Id and all fingerprint data. ✔
- Add non-default constructors to Gift and address all warnings (see BlogEngine's Post.cs file). ✔
- Add any missing entity framework CRUD unit tests for `Gift`. ✔
- Implement an Automapper Configuration between two `Gift` objects. ✔
  - Do not map the Id ✔
- Create services for (do not map primary key Id for all entities):
  - Groups ✔
  - Users ✔
  - Gifts ✔
  - When retrieving the Gift object from the service, User should be pre-populated (included) ✔
- Refactor out SampleData (see BlogEngine's SampleData.cs) ✔

## Extra Credit

- Write exceptional code that is clean, conforms to all coding guidelines, is well refactored, and easy to understand. ✔

## Project

| ProjectName | Nullable | Solutions | Analyzers | Unexpected Disabled Warnings | Waning Level |
| -- | -- | -- | -- | -- | -- |
| SecretSanta.Api | Enable | SecretSanta.sln |  |  | 4 |
| SecretSanta.Business | Enable | SecretSanta.sln |  |  | 4 |
| SecretSanta.Data | Enable | SecretSanta.sln |  |  | 4 |
| SecretSanta.Web | Enable | SecretSanta.sln |  |  | 4 |
| SecretSanta.Api.Tests | Enable | SecretSanta.sln |  |  | 4 |
| SecretSanta.Business.Tests | Enable | SecretSanta.sln |  |  | 4 |
| SecretSanta.Data.Tests | Enable | SecretSanta.sln |  |  | 4 |
| SecretSanta.Web.Tests | Enable | SecretSanta.sln |  |  | 4 |


| Project Analysis |   |
| -----------------| - |
| Build/Compile | (Build succeeded: 0 Error(s), 9 Warning(s)) |
| Disabled Warnings **** | **✔**
| Nullable Enabled: | **✔** |
| Appropriately handle all warnings | **❌** |
| Test Run Succeeded: Total tests=42, Passed=42, Failed=0, Skipped=0 |

**Pull request closed without merging**